feat: improve the image management experience with template builder (#27018)
coder/coder: docs/admin/templates/managing-templates/image-management.md

# Image Management

While Coder provides example
[container images](https://github.com/coder/images) for
workspaces, it's often best to create custom images that match the needs of
your users. This document serves as a guide to operational maturity with some
best practices around managing workspace images for Coder.

> 1. Create a minimal base image

Reviewer: ```suggestion
best practices around managing workspace images for Coder.

After following this tutorial, you'll accomplish the following:

1. Create a minimal base image.
```

I thought it was odd that this list was just here with no preface.

> 2. Create golden image(s) with standard tooling

Reviewer: ```suggestion
2. Create golden images with standard tooling.
```

3. Create project-specific images for common use cases

> 4. Let developers customize their own environment

Reviewer: ```suggestion
3. Create project-specific images for common use cases.
4. Let developers customize their own environment.
```

An image is just one of the many properties defined within the template.
Templates can pull images from a public image registry (e.g. Docker Hub) or an
internal one, thanks to Terraform. Each image reference below can be dropped
directly into a template's image variable or parameter.

## Create a minimal base image

While you may not use this directly in Coder templates, it's useful to have a

> minimal base image: a small image that contains only the necessary

Reviewer: ```suggestion
minimal base image as a small image that contains only the necessary
```

  `docker`, `bash`, `jq`, and/or internal tooling
- Consider creating (and starting the container with) a non-root user

Examples:

- [`codercom/example-minimal:ubuntu`](https://github.com/coder/images/tree/main/images/minimal): only the necessary dependencies for a Coder workspace to bootstrap
- [`codercom/example-base:ubuntu`](https://github.com/coder/images/tree/main/images/base): a slightly more padded starting point with common utilities preinstalled

> ## Create golden image(s) with standard tooling

Reviewer: ```suggestion
## Create golden images with standard tooling
```

Examples:

- [`codercom/example-universal:ubuntu`](https://github.com/coder/images/tree/main/images/universal): a catch-all image with many languages and tools preinstalled. Runs as the `coder` user, so it works with Coder templates out of the box.
- [`mcr.microsoft.com/devcontainers/universal`](https://github.com/devcontainers/images/tree/main/src/universal): the Universal Dev Containers image

## Create project-specific images for common use cases

Beyond golden images, create images scoped to a specific project, language, or

> use case (e.g. a Go backend, a Node.js frontend, or a data science stack).

Reviewer: ```suggestion
use case (e.g., a Go backend, a Node.js frontend, or a data science stack).
```

These images stay smaller and faster to pull than a kitchen-sink image, and

> give each team exactly the tooling they need.

Reviewer: ```suggestion
These images stay smaller and faster to pull than one larger image that tries to install all dependencies.
```

People whose first language isn't English may not understand the "kitchen-sink" idiom.

Examples:

- [`codercom/example-golang:ubuntu`](https://github.com/coder/images/tree/main/images/golang),
  [`codercom/example-java:ubuntu`](https://github.com/coder/images/tree/main/images/java),
  [`codercom/example-node:ubuntu`](https://github.com/coder/images/tree/main/images/node):
  Coder's example language-specific images, all running as the `coder` user.

>   See [coder/images](https://github.com/coder/images) for the full list.

Reviewer: ```suggestion
  Refer to [coder/images](https://github.com/coder/images) for the full list.
```

- [`codercom/oss-dogfood:latest`](https://github.com/coder/coder/tree/main/dogfood):
  the image Coder's own engineers use to develop Coder. A good reference for a
  project-specific image tailored to a specific team or monorepo setup.

## Let developers customize their own environment

Even with well-scoped images, developers will often need tooling that is
specific to their project or personal workflow. Instead of maintaining an
image for every combination, let developers layer their own customizations

> on top of a smaller image:

Reviewer: ```suggestion
image for every combination, developers can layer their own customizations
on top of a smaller image:
```

- [Dev Containers](https://containers.dev): developers define their project's
  dependencies in a `devcontainer.json` in their Git repository, and Coder

>   builds the environment on top of your base image. See

Reviewer: ```suggestion
  builds the environment on top of your base image. Visit
```

  [configure a template for Dev Containers](../../integrations/devcontainers/integration.md).
- [mise](https://mise.jdx.dev/): developers install and pin language runtimes

>   and CLI tools at workspace startup without rebuilding the image. See the

Reviewer: ```suggestion
  and CLI tools at workspace startup without rebuilding the image. Visit the
```
